Add documentation for logic customization (#9325)
decidim/decidim: docs/modules/customize/pages/logic.adoc

Now, let's move our modifications from the first copy-paste example into a concern.

. Create the file `app/forms/concerns/admin_project_form_extensions.rb`, and make the desired changes in that file as follows:
[source,ruby]
----
# frozen_string_literal: true
module AdminProjectFormExtensions
  extend ActiveSupport::Concern
    included do
      validates :budget_amount, numericality: { less_than_or_equal_to: :maximum_budget }
    end
  private
    def maximum_budget
      (budget[:total_budget] * 0.5).round
    end
end
----
[start=2]
. After the concern is created, we need to apply it to the correct class. We can do this in the `to_prepare` hook in `config/application.rb`:
[source,ruby]
----
  config.to_prepare do
    Decidim::Budgets::Admin::ProjectForm.include(Decidim::Budgets::Admin::ProjectForm.include(AdminProjectFormExtensions)
  end
----

> This applies to almost all classes you want to change in Decidim but controllers are a bit special case due to order of how things are loaded in the Decidim boot process. If you want to add any changes to controllers, you need to wrap the `config.to_prepare` block within an initializer that is run at the correct phase of the boot process as follows:

Reviewer: I just realized this actually applies to the helpers as well when they are included in controllers, so I would request still a small change here:
```suggestion
This applies to almost all classes you want to change in Decidim but controllers and helpers are special cases due to order of how things are loaded in the Decidim boot process. If you want to add any changes to controllers or helpers, you need to wrap the `config.to_prepare` block within an initializer that is run at the correct phase of the boot process as follows:
```

> = Change logic in decdidim-core

Author: As @ahukkanen suggested, the title changed to "Decidim core logic within your application"

There are occasions you find yourself in a situation that you need to change Decidim core logic within your application. There are two ways to accomplish this goal, and we are going through these options here.
As an example, we will add a new validation rule to the Admin panel, where Admin can not add projects with budgets more than half the total budget. Once we applied such limitation, admins should not be able to add a project with the cost of 5,001 euros to a budget with total of 10,000 euros.

> [.maximum-budget-rule]

Author: I used this method based on the following explanation (which you have provided also)
https://docs.asciidoctor.org/asciidoc/latest/macros/images/#ex-attributes

![Block image macro](https://user-images.githubusercontent.com/104360479/170019251-eb35218d-350c-4749-9777-8c7f01f50e78.png)

Concerns should be placed into the `concerns` folder within the top-level folder where you want to apply these concerns. In this case, we are adding a concern for the forms classes, so we should place the concerns related to forms to `app/forms/concerns`. If you do not have this folder available, create it first `mkdir -p app/forms/concerns`.

Now, let's move our modifications from the first copy-paste example into a concern.

Create the file `app/forms/concerns/admin_project_form_extensions.rb`, and make the desired changes in that file as follows:

[source,ruby]
----
# frozen_string_literal: true
module AdminProjectFormExtensions
  extend ActiveSupport::Concern
    included do
      validates :budget_amount, numericality: { less_than_or_equal_to: :maximum_budget }
    end
  private
    def maximum_budget
      (budget[:total_budget] * 0.5).round
    end
end
----
After the concern is created, we need to apply it to the correct class. We can do this in the `to_prepare` hook in `config/application.rb`:

[source,ruby]
----
config.to_prepare do
  Decidim::Budgets::Admin::ProjectForm.include(Decidim::Budgets::Admin::ProjectForm.include(AdminProjectFormExtensions)
end
----

> 

Reviewer: Let's also add here another note if you want to change anything in the controllers:

```suggestion
This applies to almost all classes you want to change in Decidim but controllers are a bit special case due to order of how things are loaded in the Decidim boot process. If you want to add any changes to controllers, you need to wrap the `config.to_prepare` block within an initializer that is run at the correct phase of the boot process as follows:

[source,ruby]
----
initializer "customizations", after: "decidim.action_controller" do
  config.to_prepare do
    # Add the controller customizations here
  end
end
----

```

[start=3]

> Restart the server to apply these changes, as we changed the configuration of the application.

Author: IT seems that when there are blocks of codes in the previous ordered list, it can't render the next ordering list, so I have to add the numbering here

[#explanation]
== Extra note

The following step would help you find original source code:

* Because we are changing something in the budgets component, we expect to find this file from the decidim-budgets module
* To find where the gem is located, run the following command within your application:

[source,ruby]
----
bundle show decidim-budgets
----

* Look at the URL in the view which you want to change. This should generally provide you a hint where to find the correct file to change if you are already familiar with the Rails framework to begin with.
* As we are changing a validation rule, we expect these kinds of rules to be defined in the form objects which control the different forms in Decidim (so look into the app/forms folder within decidim-budgets).

> 

Reviewer: Can you please remove the extra whitespace from the end of the document?
